sessionctx/variable/session.go

```diff
@@ -863,6 +863,19 @@ type SessionVars struct {
 
 	// TemporaryTableData stores committed kv values for temporary table for current session.
 	TemporaryTableData kv.MemBuffer
+
+	// cached is used to optimze the object allocation.
+	cached struct {
+		curr int8
+		data [2]stmtctx.StatementContext
+	}
+}
+
+// InitStatementContext initializes a StatementContext, the object is reused to reduce allocation.
+func (s *SessionVars) InitStatementContext() *stmtctx.StatementContext {
+	s.cached.curr = (s.cached.curr + 1) % 2
```

Comment:
Maybe a simple `curr == 1 ? 0 : 1` is enough instead of a modulo. I personally prefer to replace modulo with `if curr == max_val then curr = 0`.

Reply:
Maybe you have a misconception that `%` is slow ...
But nowadays the compiler are very smart, they compile modulo operation to shift instructions.

For example, 

```
func BenchmarkXXX(b *testing.B) {
	for i:=0; i<b.N; i++ {
		f(0)
                // g(0)
	}
}

func f(id int) int {
	return (id +1) % 2
}

func g(id int) int {
	if id == 1 {
		return 0
	}
	return 1
}
```

go tool compile -S main_test.go

`f()` is compiled to those instructions:

```
	0x0000 00000 (main_test.go:11)	TEXT	"".f(SB), NOSPLIT|ABIInternal, $0-16
	0x0000 00000 (main_test.go:12)	MOVQ	"".id+8(SP), AX
	0x0005 00005 (main_test.go:12)	LEAQ	1(AX), CX
	0x0009 00009 (main_test.go:12)	SHRQ	$63, CX
	0x000d 00013 (main_test.go:12)	LEAQ	(CX)(AX*1), CX
	0x0011 00017 (main_test.go:12)	LEAQ	1(CX), CX
	0x0015 00021 (main_test.go:12)	SARQ	$1, CX
	0x0018 00024 (main_test.go:12)	SHLQ	$1, CX
	0x001b 00027 (main_test.go:12)	SUBQ	CX, AX
	0x001e 00030 (main_test.go:12)	INCQ	AX
	0x0021 00033 (main_test.go:12)	MOVQ	AX, "".~r1+16(SP)
	0x0026 00038 (main_test.go:12)	RET
```

`g()` is compiled to one jump instruction

```
"".g STEXT nosplit size=31 args=0x10 locals=0x0 funcid=0x0
	0x0000 00000 (main_test.go:15)	TEXT	"".g(SB), NOSPLIT|ABIInternal, $0-16
	0x0000 00000 (main_test.go:16)	MOVQ	"".id+8(SP), AX
	0x0005 00005 (main_test.go:16)	CMPQ	AX, $1
	0x0009 00009 (main_test.go:16)	JNE	21
	0x000b 00011 (main_test.go:17)	MOVQ	$0, "".~r1+16(SP)
	0x0014 00020 (main_test.go:17)	RET
	0x0015 00021 (main_test.go:19)	MOVQ	$1, "".~r1+16(SP)
	0x001e 00030 (main_test.go:19)	RET
```

Although `f` use more instructions, but remember the CPU instruction pipeline? it doesn't break the pipeline while `JNE` does! This cause the final result of `f()` and `g()` doesn't really differ.

Well, the performance doesn't really matter here. I just want to reveal the common misunderstanding about `%`

Then we compare the readability ...
`x = (x + 1) % 2` is just one line.

However, `x == 1 ? 0 : 1` is not available in Go, so `if` is used and it's 4 lines...
```
if x == 0 {
    return 1
} 
return 0
```

@xhebox

Reply:
OK, it is just a preference. I know it may be as fast as a condition. It would be faster if it is cmov, but whatever.